src/net_processing.cpp

```diff
@@ -1893,6 +2414,11 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
             ReadCompactSize(vRecv); // ignore tx count; assume it is 0.
         }
 
+        if (nCount == 0) {
```

Comment:
LOL I didn't submitted it till I saw email notification, not sure why github sent notification to you without me submitting the review :D

Re #1589 removal: I'm not sure if it's really helpful anymore, can't seem to reproduce the initial issue after latest backborts. Also, I think it's actually causing #1890 and thus it can be kind of dangerous to have conditions like that on mainnet.

Reply:
Hmm, I couldn't think of a change from backports that could have fixed the initial issue. While reproducing, did you try a full initial sync on mainnet? The issue should happen when headers are being downloaded and a new block is mined at the same time.

Reply:
I'm checking this logic now,  after @UdjinM6 's fixes all tests are passing,  I'll try to test your case with full initial sync.

Reply:
Ok, I see what's going on now. I assumed that the issue was downloading multiple header chains from _different_ peers but turns out that it's multiple header chains from a _single_ peer. This is a weird issue and I guess there must be smth wrong deep inside that needs fixing instead of postponing such requests, I'll have a look.

Reply:
pls see #2017

Reply:
Ok, I can't seem to find a good alternative to #1589 so far and it doesn't feel right to include 033f1cb here then. So I reverted it in my branch https://github.com/UdjinM6/dash/commit/dc90c622f81291b9d15d2e23f38450a7f5bdca3f and added a simple alternative patch https://github.com/UdjinM6/dash/commit/1f393ca4a2f9ea19becd927aee334249c7db2e8d to fix test issues. Corresponding Travis build looks good https://travis-ci.org/UdjinM6/dash/builds/359168685. Pls test.

Reply:
I would just drop 033f1cb, apply the new patch and force-push. After thatI think we can start a clean review from scratch :)

Reply:
btw @codablock , do you remember why `nCount == 0` was moved down exactly?

Reply:
I've added the fix (not tested it,  compiled only). How do you test it? I've tried to sync it from scratch in testnet and it has worked for me.(

Reply:
Well, I just dropped `blocks/` and `chainstate/` in `testnet/` and `devnet/` and it failed with `detected bad peer for initial headers sync, disconnecting peer 0` etc. Reindex and syncing with latest blocks after few hours of being offline worked fine with old patch before I deleted these folders, so this result was quite surprising to me. Also, I have another folder which was not synced for quite a while and after this failure I tried to sync with that -datadir specified and it failed too. So I started to look a bit deeper :)

Reply:
@UdjinM6 This was because otherwise we'd never get to the point where the `detected bad peer for initial headers sync` condition is checked. Starting reviewing now, maybe I can say more later.